### ISSUE TITLE
tag expression being parsed incorrectly

### ISSUE THREAD

**Issue by link89** [cucumber/cucumber#285]
Given a feature file has tags that include brackets, eg:
```Gherkin
@Case(1)
Feature: Nothing
```
When I run test case with cucumber-java with following options
```java
    @RunWith(Cucumber.class)
    @CucumberOptions(
            format = {"json:target/json-report/dw.json"}
            , features = {"classpath:features/FixedAmountWithdraw.feature"}
            , glue = {"io.cucumber.samples.dw.steps"}
            , tags = "@Case(1)"
    )
    public class AppTest {
    }
```

Then a exception is thrown:
```
java.lang.Error: Not empty
	at io.cucumber.tagexpressions.TagExpressionParser.parse(TagExpressionParser.java:70)
	at cucumber.runtime.TagPredicate.<init>(TagPredicate.java:28)
	at cucumber.runtime.Runtime.<init>(Runtime.java:69)
	at cucumber.runtime.Runtime.<init>(Runtime.java:45)
	at cucumber.runtime.Runtime.<init>(Runtime.java:41)
	at cucumber.api.junit.Cucumber.createRuntime(Cucumber.java:84)
	at cucumber.api.junit.Cucumber.<init>(Cucumber.java:63)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:104)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:86)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:33)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:250)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

It turns out that tag expression parser doesn't handle tags with brackets properly. I think tags in tag expressions should be processed as whole rather than being breakdown. This issue annoy us a lot since we have hundreds of feature files with tags like this, and they are failed to run after we upgrade cucumber-java from 1.x to 2.x

**Comment by aslakhellesoy** [comment]
The tag expression grammar uses parenthesis for complex boolean expressions, for example:

    not @foo and (@bar or @zap)

If we allowed parenthesis as part of a tag, should we also enforce that they're balanced? Would `@Case(1` be ok? Even if we said tags would have to be balanced, I think it would complicate the tag expression parser too much.

Why is it important for you to use parenthesis?

**Comment by aslakhellesoy** [comment]
@link89 if you feel strongly about it I welcome you or one of your colleagues to submit a pull request that adds support for parenthesis in tag expressions. We'd be happy to include this in the next release, but you probably have to implement it yourself.

Start here: https://github.com/cucumber/cucumber/tree/master/tag-expressions/java

**Comment by link89** [comment]
> Why is it important for you to use parenthesis?
We maintain and run test cases programmatically, and this format of tags are widely used for a long time. It would be annoy to modify our test framework and all feature files to works with this change.

> If we allowed parenthesis as part of a tag, should we also enforce that they're balanced?
I don't think so. Since parenthesis become parts of tags expressions grammar, I think a better technique for this problem is escaping. When parenthesis are used in tag names, escape by using double parenthesis. eg, 
```java
    @RunWith(Cucumber.class)
    @CucumberOptions(
            format = {"json:target/json-report/dw.json"}
            , features = {"classpath:features/FixedAmountWithdraw.feature"}
            , glue = {"io.cucumber.samples.dw.steps"}
            , tags = "Case\\(1\\)"
    )
    public class AppTest {
    }
```
And things would become much easy for us.

> if you feel strongly about it I welcome you or one of your colleagues to submit a pull request that adds support for parenthesis in tag expressions. 
If you would agree with this solution, we would like to implement this feature.

**Comment by aslakhellesoy** [comment]
Ok, that sounds like a pragmatic solution to me. All 3 implementations (Java, Ruby, JavaScript) would have be updated before we can make a new release of the library (we release them in parallel, with the same version numbers).

Thanks!

**Comment by link89** [comment]
#286 here is the PR
I am not familiar with ruby so I use a regex trick for this feature in ruby's implementations. Thought passing all test cases, it will fail in some special cases when tags name contain backslash.

**Comment by aslakhellesoy** [comment]
Fixed by #286
